machine_test.go

```diff
@@ -1772,6 +1821,261 @@ func TestCreateSnapshot(t *testing.T) {
+	}{
+		{
+			name: "TestLoadSnapshot without create",
+			createSnapshot: func(ctx context.Context, machineLogger *logrus.Logger, socketPath, memPath, snapPath string) {
+
+			},
+
+			loadSnapshot: func(ctx context.Context, machineLogger *logrus.Logger, socketPath, memPath, snapPath string) {
+				cfg := createValidConfig(t, socketPath+".load")
+				m, err := NewMachine(ctx, cfg, func(m *Machine) {
+					// Rewriting m.cmd partially wouldn't work since Cmd has
+					// some unexported members
+					args := m.cmd.Args[1:]
+					m.cmd = exec.Command(getFirecrackerBinaryPath(), args...)
+				}, WithLogger(logrus.NewEntry(machineLogger)))
+				require.NoError(t, err)
+
+				err = m.Start(ctx, WithSnapshot(memPath, snapPath))
+				require.Error(t, err)
+			},
+		},
+		{
+			name: "TestLoadSnapshot and check contents (via ssh)",
+			createSnapshot: func(ctx context.Context, machineLogger *logrus.Logger, socketPath, memPath, snapPath string) {
+				cniConfPath := fmt.Sprintf("%s/%s.conflist", cniConfDir, networkName)
+				err := writeCNIConfWithHostLocalSubnet(cniConfPath, networkName, subnet)
+				require.NoError(t, err)
+				defer os.Remove(cniConfPath)
+
+				networkInterface := NetworkInterface{
+					CNIConfiguration: &CNIConfiguration{
+						NetworkName: networkName,
+						IfName:      ifName,
+						ConfDir:     cniConfDir,
+						BinPath:     cniBinPath,
+						VMIfName:    "eth0",
+					},
+				}
+				cfg := createValidConfig(t, fmt.Sprintf("%s.create", socketPath),
+					withRootDrive(rootfsPath),
+					withNetworkInterface(networkInterface),
+				)
+
+				cmd := VMCommandBuilder{}.WithSocketPath(fmt.Sprintf("%s.create", socketPath)).WithBin(getFirecrackerBinaryPath()).Build(ctx)
+
+				m, err := NewMachine(ctx, cfg, WithProcessRunner(cmd))
+				require.NoError(t, err)
+
+				err = m.Start(ctx)
+				require.NoError(t, err)
+				defer m.StopVMM()
+				defer func() {
+					if err := m.Shutdown(ctx); err != nil {
+						t.Logf("error on vm shutdown: %v", err)
+					}
+				}()
+
+				var client *ssh.Client
+
+				for tmp := 0; tmp < 20; tmp++ {
+					client, err = connectToVM(m, sshKeyPath)
+					if err != nil {
+						time.Sleep(1 * time.Second)
+					} else {
+						break
+					}
+				}
+				require.NoError(t, err)
+				defer client.Close()
+
+				ipToRestore = m.Cfg.NetworkInterfaces.staticIPInterface().StaticConfiguration.IPConfiguration.IPAddr.IP.String()
+
+				session, err := client.NewSession()
+				require.NoError(t, err)
+				defer session.Close()
+
+				err = session.Start(`sleep 422`) // Should be asynchronous
+				require.NoError(t, err)
+
+				err = m.PauseVM(ctx)
+				require.NoError(t, err)
+
+				err = m.CreateSnapshot(ctx, memPath, snapPath)
+				require.NoError(t, err)
+			},
+
+			loadSnapshot: func(ctx context.Context, machineLogger *logrus.Logger, socketPath, memPath, snapPath string) {
+				time.Sleep(10 * time.Second) // Let the system free IP
```

Comment:
@sondavidb, I did some digging and found a more deterministic way to have this work. So effectively we need to create a temporary cache directory and use it in the network interface configuration for both save and load snapshot.

```go
				networkInterface := NetworkInterface{
					CNIConfiguration: &CNIConfiguration{
						NetworkName: networkName,
						IfName:      ifName,
						ConfDir:     cniConfDir,
						CacheDir: cniCacheDir,
						BinPath:     cniBinPath,
						Args:        [][2]string{{"IP", ipToRestore + networkMask}},
						VMIfName:    "eth0",
					},
				}
```
Then we will need to monitor this directory for the IP to be freed before continuing to load the snapshot.